### PR TITLE
Add spack_short_version variable in path substitutions

### PR DESCRIPTION
**Pull request by Jordan474**
Add `$spack_short_version` variable in Spack's path variable substitutions.

I am using these variables because I actively work with multiple Spack instances of different versions simultaneously. I have found that isolating some Spack directories can be very helpful:

```yaml
config:
  # Facilitate the use of multiple Spack versions:
  # - Avoid systematically rebuilding recipe/provider cache
  misc_cache: $user_cache_path/cache$spack_short_version
  # - Avoid unnecessary contention on build stage
  build_stage: [$tempdir/$user/spack$spack_short_version-stage]
```
```yaml
bootstrap:
  # Facilitate the use of multiple Spack versions:
  # - Isolate bootstrap to avoid store db upgrades and re-indexing on downgrades
  root: $user_cache_path/bootstrap$spack_short_version
```

(I also add `$os-$target_family` on infrastructures with overly shared FS)

### CODE REVIEW

**Review by alalazo — COMMENTED** [review]
I think it's a very good idea. Was wondering (and discussing with @haampie) if we can get by with two variables:
- `short_version`, e.g.  0.22
- `full_version`, e.g. 0.22.2

What do you think?

**Comment by Jordan474** [comment]
I didn't want to make any semantic assumptions about stopping after 1, 2, or 3 components. I even hesitated between using `x` or `first` instead of `major`. For instance, I didn't want to ask myself the question of `short_version` after Spack v1: should it be `1` or `1.0`? So, I left it up to the user to decide.

Regarding naming, I have no preference between `$version_X`, `$X_version`, and `$spack_X`.

Either options works for me, so I'll let you have the last word.

**Comment by Jordan474** [comment]
I have another PR that *maybe* could use similar semantic and naming for consistence: https://github.com/spack/spack/pull/38410

**Comment by haampie** [comment]
> For instance, I didn't want to ask myself the question of short_version after Spack v1: should it be 1 or 1.0? So, I left it up to the user to decide.

I do, it's `1.0` :p

I would initially only add `spack_short_version`, and leave `full_version` out, cause both cache and config should be compatible, and I don't see where else this could possibly be useful. I'm generally a fan of adding fewer not more things to spack ;)

**Comment by Jordan474** [comment]
I changed everything to implement only `$spack_short_version`.

**Comment by alalazo** [comment]
Fwiw I asked other maintainers if they're fine also changing the defaults in `etc` for the next release, once this is in. :crossed_fingers:

**Review by alalazo — APPROVED** [review]

**Pull request merged**
